src/openforms/js/components/admin/form_design/data.js

```diff
@@ -196,10 +211,32 @@ const savePriceRules = async (formUrl, csrftoken, priceRules, priceRulesToDelete
   return createdRules;
 };
 
-const createOrUpdateFormVariables = async (formUrl, csrftoken, variables) => {
+const createOrUpdateFormVariables = async (
+  formUrl,
+  csrftoken,
+  variables,
+  stateformSteps,
+  updatedFormSteps
+) => {
   const endPoint = `${formUrl}/variables`;
   const formVariables = variables.map(variable => {
-    return {...variable, form: formUrl};
+    // There are 2 cases here:
+    // 1. The variable was added to an existing form definition (i.e. which already has a URL). So variable.formDefinition
+    //   is the URL of the formDefinition.
+    // 2. The variable was added to a new form definition (i.e. which didn't have a URL). So variable.formDefinition
+    //   is the _generatedId of the formDefinition.
+    let formDefinitionUrl = variable.formDefinition;
+    try {
+      new URL(variable.formDefinition);
+    } catch (e) {
+      // Retrieve the URL of the definition from the formSteps which have already been saved
+      stateformSteps.map((step, index) => {
+        if (step._generatedId === variable.formDefinition) {
+          formDefinitionUrl = updatedFormSteps[index].formDefinition;
+        }
+      });
+    }
+    return {...variable, form: formUrl, formDefinition: formDefinitionUrl};
```

Comment:
Is this something that can be done in the reducer rather than in the data layer possibly? E.g. when the form definition is created, dispatch an action to update the internal react state and do this merging/joining at that time rather than deferring it? I'm not sure if that would lead to simpler code though.

Reply:
I am not sure, because this function `createOrUpdateFormVariables` is called from the `onSubmit` function which also makes the calls to save the `formDefinition`s. When the `createOrUpdateFormVariables` is called, the react state has not been updated yet with the results of the previous calls (so `stateFormSteps` doesn't contain the new `formDefinition` URLs yet) :thinking:

Reply:
yeah I think we'd have to revisit that with immer's `produce` separately maybe :thinking: